File: VantagePro2Driver/src/vantagepro2/VantagePro2Driver.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.IO.Ports;
using System.Threading;
using System.Diagnostics;

namespace VantagePro2 {
    /// <summary>
    /// This is the main loop that contains the business logic for talking to the Vantage Pro console.
    /// </summary>
    public class VantagePro2Driver : VantagePro2.WeatherSenseSocket.HistoricalReader, VantagePro2.VantagePro2Station.Callback {
        private VantagePro2Station station;
        private WeatherSenseSocket socket;
        private ArchiveManager archiveManager = new ArchiveManager("C:\\WeatherSense\\2.3\\archive\\archive.vp2");
        private bool firstPass = true;
        private bool firstCurrentWeather = true;
        private bool exit = false;
        private Thread thread;
        private int nextRecord = -1;
        private int previousNextRecord = -1;
        private readonly int LOOP_PACKET_CYCLES = 12;
        private ParametersMessage parameters;
        private DateTime lastPacketTime = new DateTime(2000, 6, 22, 17, 0, 0);
        private DateTime consoleTimeSetTime = new DateTime(2000, 1, 1, 0, 0, 0);
        private DateTime sensorStationSendTime = new DateTime(2000, 1, 1, 0, 0, 0);

        /// <summary>
        /// Constructor.
        /// </summary>
        /// <param name="port">The serial port to open</param>
        /// <param name="baudRate">The baud rate at which to read from the port</param>
        public VantagePro2Driver(String port, int baudRate) {
            Debug.AutoFlush = true;
           
            station = new VantagePro2Station(port, baudRate, this);
            socket = new WeatherSenseSocket(this);
        }

        /// <summary>
        /// Called by the socket class when the response is received from the WeatherSense collector. The response contains
        /// the date of the newest archive record in the database. This is the basis for the first dump after command.
        /// </summary>
        /// <param name="time">The time of the newest record in the database</param>
        public void newestHistoryRecordTime(DateTime time) {
            lastPacketTime = time;
            nextRecord = -1;
        }

        /// <summary>
        /// Called by the socket class when it connects to the WeatherSense collector.
        /// </summary>
        public void connected() {
            firstPass = true;
            firstCurrentWeather = true;
        }

        /// <summary>
        /// Entry point for the main loop
        /// </summary>
        public void start() {

            if (!station.Open()) {
                //System.Diagnostics.EventLog.
                Debug.Print("Failed to open weather station");
                Environment.Exit(1);
            }
            else {
                Debug.Print("Port is open");
            }

            if (!station.Wakeup()) {
                Debug.Print("Failed to wake up weather station");
                Environment.Exit(1);
            }
            else {
                Debug.Print("Weather Station is awake");
            }

            //
            // Get the information from the console that we will need for future calculations
            //
            if (!station.retrieveSetupValues() || !station.determineArchivePeriod() || !station.sensorStationInfo()) {
                Debug.Print("Failed to retrieve configuration information");
                Environment.Exit(2);
            }

            //
            // Kick off our own thread
            //
            thread = new Thread(new ThreadStart(this.run));
            thread.Start();

            //
            // Once the thread is alive this thread can exit
            //
            while (!thread.IsAlive) ;
        }

        /// <summary>
        /// Stop the thread and close all of the resources.
        /// </summary>
        public void stop() {
            exit = true;
            thread.Join();
            station.Close();
            socket.disconnect();
        }

        /// <summary>
        /// Callback used to process a current weather record.
        /// </summary>
        /// <param name="cw">The current weather</param>
        /// <returns>True if the loop packet loop should continue, false if it should exit and try an achive dump</returns>
        public bool processCurrentWeather(CurrentWeather cw) {
            //
            // First time through figure out what sensors are connected to the weather station using the values in the
            // LOOP packet
            //
            if (firstCurrentWeather) {
                firstCurrentWeather = false;
                String sensorMessage = Sensor.formatMessage(station.Sensors);
                socket.sendData(sensorMessage);
            }

            String currentWeatherMessage = cw.formatMessage();
            nextRecord = cw.NextPacket;
            socket.sendData(currentWeatherMessage);
            Debug.Print("Previous Next Record: {0}  Next Record: {1}", previousNextRecord, nextRecord);
            return previousNextRecord == nextRecord;
        }

        /// <summary>
        /// Callback to process a page of archive packets.
        /// </summary>
        /// <param name="page">A list of archive packets that make up the page</param>
        /// <returns>True if all of the packets were sent to the WeatherSense collector</returns>
        public bool processArchivePage(List<ArchivePacket> page) {
            archiveManager.addPackets(page);
            if (firstPass)
                return true;

            foreach (ArchivePacket packet in page) {
                //
                // If the page has a null entry then this is the last page of the DMPAFT command.
                // Set the previousNextRecord to the nextRecord so that we know when the next record
                // changes again and we need to do another DMPAFT.
                //
                if (packet == null) {
                    previousNextRecord = nextRecord;
                }
                else {
                    DateTime now = DateTime.Now;
                    TimeSpan age = now.Subtract(packet.Time);
                    if (age.TotalHours < 1) {
                        int maxPackets = (int)((station.ArchivePeriod * 60F) / ((41F + 1F - 1F) / 16F));
                        int actualPackets = packet.WindSampleCount;
                        int issReception = (actualPackets * 100) / maxPackets;
                        if (issReception > 100)
                            issReception = 100;

                        Debug.Print("IIS Reception for archive interval ending at {0} is {1}. Max Packets = {2}, Actual Packets - {3}",
                                     packet.Time, issReception, maxPackets, actualPackets);

                        List<SensorStation> sensorStations = station.SensorStations;
                        foreach (SensorStation sensorStation in sensorStations) {
                            if (sensorStation.Type == SensorStation.SensorStationType.INTEGRATED_SENSOR_STATION)
                                sensorStation.LinkQuality = issReception;
                        }

                        String ssMessage = SensorStation.formatSensorStationStatusMessage(sensorStations, packet.Time);
                        socket.sendData(ssMessage);
                    }

                    String message = packet.formatMessage();
                    Debug.Print("=== Achive === " + packet.Time + " =============");
                    if (socket.sendData(message))
                        lastPacketTime = packet.Time;
                    else
                        return false;
                }
            }

            return true;
        }

        /// <summary>
        /// Thread entry point that is the main loop of the Vantage Pro driver
        /// </summary>
        public void run() {
            DateTime time = station.GetTime();

            Debug.Print("Station Time: {0}", time);
            DateTime archiveTime = archiveManager.findNewestPacketTime();
            station.dumpAfter(archiveTime);
            
            while (!exit) {

                try {
                    station.Wakeup();

                    //
                    // First time through get various parameters about the weather station and send them to the collector
                    //
                    if (firstPass) {
                        parameters = station.getParameters();
                        if (parameters != null) {
                            String parametersMessage = parameters.formatMessage();
                            Debug.Print(parametersMessage);
                            socket.sendData(parametersMessage);
                            firstPass = false;
                        }

                        String message = SensorStation.formatSensorStationMessage(station.SensorStations);
                        socket.sendData(message);
                    }

                    //
                    // If it has been more than a day since the time was set, set the time
                    //
                    DateTime now = DateTime.Now;
                    if (consoleTimeSetTime.AddDays(1.0).CompareTo(now) <= 0) {
                        if (station.setConsoleTime())
                            consoleTimeSetTime = now;                     
                    }

                    //
                    // Get the current weather values for about a minute
                    //
                    station.currentValuesLoop(LOOP_PACKET_CYCLES);

                    //
                    // If the LOOP packet data indicates that a new archive packet is available
                    // go get it.
                    //
                    if (previousNextRecord != nextRecord && socket.isConnected()) {
                        station.dumpAfter(lastPacketTime);
                    }
                }
                catch (Exception e) {
                    Debug.Print("Caught exception: {0}\n{1}", e.Message, e.StackTrace);     
                } 
            }
        }
    }
}

File: VantagePro2Driver/VP2++/src/VP2/VantagePro2Driver.cpp
/* 
 * Copyright (C) 2022 Bruce Beisel
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
#include <stdlib.h>
#include <time.h>
#include <iostream>
#include <vector>
#include <atomic>
#include "CurrentWeather.h"
#include "CurrentWeatherPublisher.h"
#include "HiLowPacket.h"
#include "SensorStation.h"
#include "VP2Logger.h"
#include "VP2Decoder.h"
#include "VP2Constants.h"
#include "Alarm.h"
#include "VantagePro2Driver.h"

using namespace std;
extern atomic_bool signalCaught;

namespace vp2 {

////////////////////////////////////////////////////////////////////////////////
////////////////////////////////////////////////////////////////////////////////
VantagePro2Driver::VantagePro2Driver(ArchiveManager & archiveManager, CurrentWeatherPublisher & cwp, VantagePro2Station & station, EventManager & evtMgr) :
                                                                station(station),
                                                                currentWeatherPublisher(cwp),
                                                                archiveManager(archiveManager),
                                                                eventManager(evtMgr),
                                                                exitLoop(false),
                                                                receivedFirstLoopPacket(false),
                                                                nextRecord(-1),
                                                                previousNextRecord(-1),
                                                                lastArchivePacketTime(0),
                                                                log(VP2Logger::getLogger("VantagePro2Driver")),
                                                                sensorStationSendTime(0) {
    //
    // Indicate the the console time needs to be set in the near future. 
    // We do not want the console time to be set immediately in case the computer has just started and
    // has not had a chance to synchronize its time with the Internet. This is most important with
    // computers like the Raspberry Pi.
    //
    consoleTimeSetTime = time(0) - TIME_SET_INTERVAL + (1 * SECONDS_PER_HOUR);
}

////////////////////////////////////////////////////////////////////////////////
////////////////////////////////////////////////////////////////////////////////
VantagePro2Driver::~VantagePro2Driver() {
}

////////////////////////////////////////////////////////////////////////////////
////////////////////////////////////////////////////////////////////////////////
bool
VantagePro2Driver::initialize() {
    log.log(VP2Logger::VP2_INFO) << "Initializing..." << endl;

    station.setCallback(*this);

    if (!station.openStation()) {
        log.log(VP2Logger::VP2_ERROR) << "Failed to open weather station" << endl;
        return false;
    }
    else {
        log.log(VP2Logger::VP2_INFO) << "Port is open" << endl;
    }

    if (!station.wakeupStation()) {
        log.log(VP2Logger::VP2_ERROR) << "Failed to wake up weather station" << endl;
        return false;
    }
    else {
        log.log(VP2Logger::VP2_INFO) << "Weather Station is awake" << endl;
    }

    VP2Decoder::setRainCollectorSize(station.getRainCollectorSize()); // TBD, this should come from VantagePro2Configuration class

    /*
    if (!station.retrieveConfigurationParameters()) {
        log.log(VP2Logger::VP2_ERROR) << "Failed to retrieve configuration parameters" << endl;
        return false;
    }
    */

    //
    // Get one current weather record so that the sensors are detected
    //
    for (int i = 0; i < INITIAL_LOOP_PACKET_RETRIES && !receivedFirstLoopPacket; i++) {
        station.currentValuesLoop(1);
    }

    if (!receivedFirstLoopPacket) {
        log.log(VP2Logger::VP2_ERROR) << "Failed to receive a LOOP packet needed to determine current sensor suite" << endl;
        return false;
    }

    if (!archiveManager.synchronizeArchive()) {
        log.log(VP2Logger::VP2_ERROR) << "Failed to read the archive during initialization" << endl;
        return false;
    }

    AlarmManager::getInstance().initialize();

    if (!station.wakeupStation()) {
        log.log(VP2Logger::VP2_ERROR) << "Failed to wake up console after initialization" << endl;
        return false;
    }

    log.log(VP2Logger::VP2_INFO) << "Initialization complete." << endl;

    return true;
}

////////////////////////////////////////////////////////////////////////////////
////////////////////////////////////////////////////////////////////////////////
void
VantagePro2Driver::stop() {
    exitLoop = true;
    station.closeStation();
}

////////////////////////////////////////////////////////////////////////////////
////////////////////////////////////////////////////////////////////////////////
bool
VantagePro2Driver::processCurrentWeather(const CurrentWeather & cw) {
    nextRecord = cw.getNextPacket();

    if (receivedFirstLoopPacket)
        currentWeatherPublisher.sendCurrentWeather(cw);

    log.log(VP2Logger::VP2_DEBUG1) << "Previous Next Record: " << previousNextRecord << " Next Record: " << nextRecord << endl;
    receivedFirstLoopPacket = true;

    bool sc = signalCaught.load();
    bool em = eventManager.isEventAvailable();
    bool nr = previousNextRecord != nextRecord;
    bool stopCurrentWeatherLoop = sc || em || nr;

    log.log(VP2Logger::VP2_DEBUG1) << "Stop current weather loop: " << stopCurrentWeatherLoop
                                   << " Signal: " << sc << " Event: " << em << " Next Record: " << nr << endl;

    return signalCaught.load() || eventManager.isEventAvailable() || previousNextRecord != nextRecord;
}

/*
////////////////////////////////////////////////////////////////////////////////
////////////////////////////////////////////////////////////////////////////////
bool
VantagePro2Driver::processArchive(const vector<ArchivePacket> & archive) {

    log.log(VP2Logger::VP2_DEBUG1) << "Processing " << archive.size() << " archive packets" << endl;

    for (vector<ArchivePacket>::const_iterator it = archive.begin(); it != archive.end(); ++it) {
        DateTime now = time(0);
        DateTime age = now - it->getDateTime();
        if (age < SECONDS_PER_HOUR) {
            int maxPackets = static_cast<int>(((static_cast<float>(station.getArchivePeriod()) * 60.0F) / ((41.0F + 1.0F - 1.0F) / 16.0F)));
            int actualPackets = it->getWindSampleCount();
            int issReception = (actualPackets * 100) / maxPackets;
            if (issReception > 100)
                issReception = 100;

            log.log(VP2Logger::VP2_DEBUG2) << "IIS Reception for archive interval ending at " << it->getDateTime()
                                           << " is " << issReception
                                           << ". Max Packets = " << maxPackets
                                           << ", Actual Packets - " << actualPackets << endl;

            vector<SensorStation> sensorStations = station.getSensorStations();
            for (vector<SensorStation>::iterator it2 = sensorStations.begin(); it2 != sensorStations.end(); ++it2) {
                if (it2->getSensorStationType() == SensorStation::INTEGRATED_SENSOR_STATION)
                    it2->setLinkQuality(issReception);
            }

            string ssMessage = SensorStation::formatSensorStationStatusMessage(sensorStations, it->getDateTime());
            socket.sendData(ssMessage);
        }

        string message = it->formatMessage();
        log.log(VP2Logger::VP2_INFO) << "=== Archive === " << Weather::formatDateTime(it->getDateTime()) << " =============" << endl;
        socket.sendData(message);
    }

    return true;
}
*/

////////////////////////////////////////////////////////////////////////////////
////////////////////////////////////////////////////////////////////////////////
bool
VantagePro2Driver::reopenStation() {
    station.closeStation();
    bool success = station.openStation();

    if (!success)
        log.log(VP2Logger::VP2_ERROR) << "Failed to reopen weather station" << endl;

    return success;
}

////////////////////////////////////////////////////////////////////////////////
////////////////////////////////////////////////////////////////////////////////
void
VantagePro2Driver::mainLoop() {
    vector<ArchivePacket> list;
    list.reserve(VP2Constants::NUM_ARCHIVE_RECORDS);

    while (!exitLoop) {
        try {
            //
            // If the weather station could not be woken, then close and open
            // the console. It has been observed that on a rare occasion the console
            // never wakes up. Only restarting this driver fixes the issue. Reopening
            // the serial port will hopefully fix this issue. It is hoped that if
            // this does not work, it will cause the health monitor to restart the
            // driver, also fixing the issue.
            //
            if (!station.wakeupStation()) {
                reopenStation();
                continue;
            }

            DateTime consoleTime;
            if (station.retrieveConsoleTime(consoleTime))
                log.log(VP2Logger::VP2_INFO) << "Station Time: " << Weather::formatDateTime(consoleTime) << endl;
            else
                log.log(VP2Logger::VP2_INFO) << "Station Time retrieval failed" << endl;

            //
            // If it has been a while since the time was set, set the time
            //
            DateTime now = time(0);
            if (consoleTimeSetTime + TIME_SET_INTERVAL < now) {
                if (!station.updateConsoleTime())
                    log.log(VP2Logger::VP2_ERROR) << "Failed to set station time " << endl;

                consoleTimeSetTime = now;
            }

            //
            // Get the high/low values
            //
            HiLowPacket packet;
            if (station.retrieveHiLowValues(packet))
                cout << "Got hi/low packet" << endl;

            //
            // Get the current weather values for about a minute
            //
            station.currentValuesLoop(LOOP_PACKET_CYCLES);

            //
            // If an asynchronous signal was caught, then exit the loop
            //
            if (signalCaught.load()) {
                exitLoop = true;
                continue;
            }

            string event;
            while (eventManager.consumeEvent(event)) {

            }

            //
            // If the LOOP packet data indicates that a new archive packet is available
            // go get it.
            //
            if (previousNextRecord != nextRecord) {
                if (archiveManager.synchronizeArchive()) {
                    ArchivePacket packet;
                    // TBD What if multiple archive packets are received?
                    archiveManager.getNewestRecord(packet);
                    log.log(VP2Logger::VP2_DEBUG1) << "Most recent archive packet time is: "
                                                   << Weather::formatDateTime(packet.getDateTime())
                                                   << " Station Reception: " << station.calculateStationReceptionPercentage(packet.getWindSampleCount()) << endl; // TBD Get the actual archive period
                    previousNextRecord = nextRecord;
                }
            }
        }
        catch (std::exception & e) {
            log.log(VP2Logger::VP2_ERROR) << "Caught exception: " << e.what() << endl;     
        } 
    }
}
}
